File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/PlanRewriteRuleResult.java
package com.foundgine.core.semantic.planning;

import java.util.*;

/** Auditable result of applying one rewrite rule. */
public record PlanRewriteRuleResult(String ruleName,SemanticPlan before,SemanticPlan after,List<String> preconditions,List<String> securityObligations,double costImpact,SecurityPreservationProof securityProof,SecurityObligationProof securityObligationProof,SemanticEquivalenceProof semanticProof,SemanticPlanAuthorizationBindingProof authorizationBindingProof,PlanOptimizationProof optimizationProof) {
    public PlanRewriteRuleResult {preconditions=List.copyOf(preconditions);securityObligations=List.copyOf(securityObligations);}
    public boolean isSatisfied(){return securityProof.isSatisfied()&&securityObligationProof.isSatisfied()&&semanticProof.isSatisfied()&&authorizationBindingProof.isSatisfied()&&(optimizationProof==null||optimizationProof.isSatisfied());}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/PlanRewriteRuleResult.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Auditable result of applying one rewrite rule.</summary>
public sealed record PlanRewriteRuleResult(
    string RuleName,
    SemanticPlan Before,
    SemanticPlan After,
    IReadOnlyList<string> Preconditions,
    IReadOnlyList<string> SecurityObligations,
    double CostImpact,
    SecurityPreservationProof SecurityProof,
    SecurityObligationProof SecurityObligationProof,
    SemanticEquivalenceProof SemanticProof,
    SemanticPlanAuthorizationBindingProof AuthorizationBindingProof,
    PlanOptimizationProof? OptimizationProof = null)
{
    public bool IsSatisfied =>
        SecurityProof.IsSatisfied &&
        SecurityObligationProof.IsSatisfied &&
        SemanticProof.IsSatisfied &&
        AuthorizationBindingProof.IsSatisfied &&
        (OptimizationProof?.IsSatisfied ?? true);
}
